File: nutrition_ai/recipe_search.py
"""
نظام الوصفات — بحث نصي مطبّع بديل عن FTS5 (قاعدة الوصفات صغيرة، القرار موثّق بالخطة)،
وزراعة أولية تهاجر فقط الوصفات الحقيقية من recipes_seed.py (لا اختراع بيانات).
"""
import difflib

from arabic_normalize import normalize
from models import db, Recipe, RecipeCategory, RecipeIngredient, RecipeStep, RecipeSubstitution

CONFIRM_THRESHOLD = 0.55


def seed_default_recipe_categories():
    from nutrition_ai.recipes_seed import CATEGORIES_SEED
    if RecipeCategory.query.first():
        return
    for c in CATEGORIES_SEED:
        db.session.add(RecipeCategory(name=c["name"], icon=c["icon"], order_index=c["order_index"]))
    db.session.commit()


def seed_default_recipes():
    """يهاجر الـ4 وصفات الحقيقية فقط من recipes_seed.py — مرة وحدة إذا الجدول فاضي."""
    seed_default_recipe_categories()

    if Recipe.query.first():
        return

    from nutrition_ai.recipes_seed import RECIPES_SEED

    categories_by_name = {c.name: c for c in RecipeCategory.query.all()}

    for r in RECIPES_SEED:
        category = categories_by_name.get(r["category"])
        if category is None:
            continue
        recipe = Recipe(
            name=r["name"], slug=r["slug"], description=r["description"],
            category_id=category.id, prep_time_min=r["prep_time_min"],
            cook_time_min=r["cook_time_min"], servings=r["servings"], difficulty=r["difficulty"],
            calories=r["calories"], protein=r["protein"], carbs=r["carbs"], fat=r["fat"],
            match_keywords="|".join(r["match_keywords"]),
        )
        db.session.add(recipe)
        db.session.flush()

        for idx, ing in enumerate(r["ingredients"]):
            db.session.add(RecipeIngredient(
                recipe_id=recipe.id, name=ing["name"], quantity=ing["quantity"],
                unit=ing["unit"], order_index=idx,
            ))
        for step_num, step in enumerate(r["steps"], start=1):
            db.session.add(RecipeStep(
                recipe_id=recipe.id, step_number=step_num,
                instruction=step["instruction"], duration=step["duration"],
                temperature=step["temperature"], tip=step["tip"], warning=step["warning"],
            ))
        for ing_name, replacement in r["substitutions"].items():
            db.session.add(RecipeSubstitution(
                recipe_id=recipe.id, ingredient_name=ing_name, replacement=replacement,
            ))

    db.session.commit()


def _substring_match(query_norm: str, field: str) -> bool:
    f_norm = normalize(field or "")
    if not f_norm:
        return False
    return query_norm in f_norm or f_norm in query_norm


def _fuzzy_match(query_words: list, field: str) -> bool:
    """تحمّل خطأ إملائي بسيط — تُستخدم فقط ضد حقول تعريفية قصيرة (اسم/كلمات مفتاحية)،
    مو ضد نص وصف حر (كلمات عامة قصيرة بالوصف كانت تعطي تطابقات فوضوية بنسبة ثقة منخفضة)."""
    f_norm = normalize(field or "")
    if not f_norm:
        return False
    for fw in f_norm.split():
        if len(fw) < 4:
            continue
        for qw in query_words:
            if qw in fw or fw in qw:
                return True
            if difflib.SequenceMatcher(None, qw, fw).ratio() >= CONFIRM_THRESHOLD:
                return True
    return False


def search_recipes(query: str = "", category_id: str | None = None):
    """يرجّع قائمة Recipe نشطة تطابق query (اسم/وصف/كلمات مفتاحية) وcategory_id إذا انعطت.
    لا يخمّن — ماكو تطابق يرجّع قائمة فاضية والواجهة تعرض Empty State."""
    q = Recipe.query.filter_by(active=True)
    if category_id:
        q = q.filter_by(category_id=category_id)
    recipes = q.order_by(Recipe.name).all()

    query = (query or "").strip()
    if not query:
        return recipes

    query_norm = normalize(query)
    query_words = [w for w in query_norm.split() if len(w) >= 3]

    matched = []
    for r in recipes:
        keywords = (r.match_keywords or "").split("|")
        if _substring_match(query_norm, r.name) or _substring_match(query_norm, r.description):
            matched.append(r)
            continue
        if any(_substring_match(query_norm, k) for k in keywords):
            matched.append(r)
            continue
        if _fuzzy_match(query_words, r.name) or any(_fuzzy_match(query_words, k) for k in keywords):
            matched.append(r)
    return matched


def find_recipe_by_text(text: str):
    """يبحث عن أفضل وصفة مطابقة لنص حر بالشات (بديل recipes_data.find_recipe القديمة)."""
    results = search_recipes(query=text)
    return results[0] if results else None


def suggest_recipes_within(remaining_calories: int, limit: int = 2) -> list:
    """وصفات حقيقية ضمن السعرات المتبقية — تُستخدم لإثراء رد ASK_RECOMMENDATION بالشات
    (إضافة فقط، لا تلمس بحث foods.sqlite الموجود). لا تخمين — فاضية لو ماكو وصفة تناسب."""
    if remaining_calories <= 0:
        return []
    fitting = [r for r in search_recipes() if r.calories <= remaining_calories]
    fitting.sort(key=lambda r: -r.calories)
    return fitting[:limit]

File: shared/nutrition-engine/recipeSearch.ts
/**
 * منفذ من nutrition_ai/recipe_search.py — بحث نصي مطبّع بديل عن FTS5 (قاعدة الوصفات صغيرة، قرار
 * موثّق بالخطة). زراعة البيانات الأولية (seed_default_recipes) غير منفَّذة هنا — عملية بذر بيانات
 * لمرة واحدة عند الإقلاع، تخص طبقة Postgres الحقيقية، لا منطق أعمال يُختبر بمعزل.
 */
import { normalize } from "./arabicNormalize.js";
import { sequenceMatcherRatio } from "./sequenceMatcher.js";
import type { Repository, RecipeRecord } from "./db/repository.js";

const CONFIRM_THRESHOLD = 0.55;

function substringMatch(queryNorm: string, field: string | null): boolean {
  const fNorm = normalize(field || "");
  if (!fNorm) return false;
  return queryNorm.includes(fNorm) || fNorm.includes(queryNorm);
}

/**
 * تحمّل خطأ إملائي بسيط — تُستخدم فقط ضد حقول تعريفية قصيرة (اسم/كلمات مفتاحية)، مو ضد نص
 * وصف حر (كلمات عامة قصيرة بالوصف كانت تعطي تطابقات فوضوية بنسبة ثقة منخفضة).
 */
function fuzzyMatch(queryWords: string[], field: string | null): boolean {
  const fNorm = normalize(field || "");
  if (!fNorm) return false;
  for (const fw of fNorm.split(/\s+/)) {
    if (fw.length < 4) continue;
    for (const qw of queryWords) {
      if (fw.includes(qw) || qw.includes(fw)) return true;
      if (sequenceMatcherRatio(qw, fw) >= CONFIRM_THRESHOLD) return true;
    }
  }
  return false;
}

/**
 * يرجّع قائمة وصفات نشطة تطابق query (اسم/وصف/كلمات مفتاحية) وcategoryId إذا انعطى. لا يخمّن —
 * ماكو تطابق يرجّع قائمة فاضية والواجهة تعرض Empty State.
 */
export async function searchRecipes(
  repo: Repository,
  query = "",
  categoryId: string | null = null,
): Promise<RecipeRecord[]> {
  const recipes = await repo.findActiveRecipes(categoryId);

  const q = (query || "").trim();
  if (!q) return recipes;

  const queryNorm = normalize(q);
  const queryWords = queryNorm.split(/\s+/).filter((w) => w.length >= 3);

  const matched: RecipeRecord[] = [];
  for (const r of recipes) {
    const keywords = (r.match_keywords || "").split("|");
    if (substringMatch(queryNorm, r.name) || substringMatch(queryNorm, r.description)) {
      matched.push(r);
      continue;
    }
    if (keywords.some((k) => substringMatch(queryNorm, k))) {
      matched.push(r);
      continue;
    }
    // بحث بالمكونات — يخلي "عندي بيض ابحثلي أكلة" أو "وصفة دجاج" يلقى وصفات تحتوي المكوّن حتى
    // لو ماكو بالاسم/الوصف/match_keywords صراحة (مثلاً وصفة اسمها "بولاو" فيها دجاج كمكوّن).
    if (r.ingredients.some((ing) => substringMatch(queryNorm, ing.name))) {
      matched.push(r);
      continue;
    }
    if (fuzzyMatch(queryWords, r.name) || keywords.some((k) => fuzzyMatch(queryWords, k))) {
      matched.push(r);
    }
  }
  return matched;
}

/** يبحث عن أفضل وصفة مطابقة لنص حر بالشات. */
export async function findRecipeByText(repo: Repository, text: string): Promise<RecipeRecord | null> {
  const results = await searchRecipes(repo, text);
  return results.length > 0 ? results[0] : null;
}

/**
 * وصفات حقيقية ضمن السعرات المتبقية — تُستخدم لإثراء رد ASK_RECOMMENDATION بالشات. لا تخمين —
 * فاضية لو ماكو وصفة تناسب.
 */
export async function suggestRecipesWithin(repo: Repository, remainingCalories: number, limit = 2): Promise<RecipeRecord[]> {
  if (remainingCalories <= 0) return [];
  const all = await searchRecipes(repo);
  const fitting = all.filter((r) => r.calories <= remainingCalories);
  fitting.sort((a, b) => b.calories - a.calories);
  return fitting.slice(0, limit);
}
